### PR TITLE
Log4J2 support in Maven builds

### PR DESCRIPTION
**Pull request by erikkemperman**
I've considered making a jhipster module for Log4J2 support, as was suggested to me elsewhere... But I am simply not sure I can commit to maintaining something like that going forward. So instead, I decided to make a little extra effort on what I had done already to my own project, in order to add it as a choice to the main generator. I do hope you'll consider adopting it!

The main limitation of this PR is that is currently only done for Maven builds, I am not sufficiently familiar with Gradle to include that... I imagine, however, that someone who knows both Maven and Gradle should find it pretty easy to remedy this, just from the changes I am proposing to `_pom.xml`.

### CODE REVIEW

**Comment by deepu105** [comment]
Really sorry @erikkemperman you should have discussed with us before doing this much work. I personally do not want to add another option and maintain it in the main generator for something which doesnt add too much value over current.

I dont see any advantage of having 2 logging options, the current log-back solution works perfectly.

Any way ill leave this open and see what @jhipster/developers think.

**Comment by jdubois** [comment]
Well I have the same problem as you @erikkemperman : I can't commit to
maintain all that code, in particular the code you have written! That's why
we did the module system.
Besides, we are going to move some part of the logback configuration
(teaser!!!!)
I'll have a look, but it would hate to lose your work : better an
unmaintained module than losing it all
Le 5 févr. 2016 8:11 AM, "Deepu K Sasidharan" <notifications@github.com> a
écrit :

**Comment by erikkemperman** [comment]
@deepu105 @jdubois 
Thanks for the quick response! Don't worry, most of this work I had already done within our own jhipster project. As I mentioned elsewhere, we are integrating with lots of other Maven modules and it is critical for us that everything uses the same logging implementation.

Actually there is not that much code changed, only `LogsResource` and `LoggerDTO` look different when this choice is made.

But all right, fair enough, I completely understand if you'd rather not have this in the main generator. In that case I will probably keep it around as a separate (but private) patch for our own internal usage!

**Comment by atomfrede** [comment]
Why not put the module under the hood of hipster-labs, such that we can maintain it together?

**Comment by erikkemperman** [comment]
@atomfrede That sounds like a good second-best option to me, if this is definitely not going into the main generator. So then, I might initialize a jhipster module and somehow transfer it to the hipster-labs account?

I would like to point out again, though, that this PR actually doesn't change the generator output significantly :-)

**Comment by erikkemperman** [comment]
Hmm, all travis builds are now failing on `require('map-stream')`? I'm pretty sure that's unrelated to this patch...

**Comment by deepu105** [comment]
yes that was from another PR I have corrected that.

Yes you can do a module under hipster-labs, i can give you access if you
are interested so that someone would maintain it even if you are not around.
we are not bothered by the code that it changes we are bothered about
adding new options coz too much is also not fun you see.

Thanks & Regards,
Deepu

**Comment by erikkemperman** [comment]
@deepu105 All right, thanks, I will consider my options. 

> we are not bothered by the code that it changes we are bothered about adding new options coz too much is also not fun you see.

That's interesting, I would have supposed it were the other way around! At least for me, having lots of choices is what I like about jhipster, which is why I imagined that an extra option with very little impact on generated code would be a plus :-)

**Comment by deepu105** [comment]
coz more options means more combinations and more issues and hence more support work and inturn more of our time, which should be better used to add features.
Of course if we need to add an option for a very hip feature we would consider it, its just that logging isnt in the top of that you see.

**Comment by erikkemperman** [comment]
Well, I am new around here and defer to your expertise... But I would be surprised if essentially the same output would cause more issues. On the other hand it would prevent any more "can we have log4j" issues :-P

But anyway, I can take a hint. Feel free to close this if you like, and again please don't feel bad -- I had already done the interesting bits earlier and only added some boilerplate stuff for this PR!

**Comment by deepu105** [comment]
sure man, you are welcome to do PRs for other stuff. Im waiting to see your gulp sourcemap changes

**Comment by erikkemperman** [comment]
I linked to my sourcemap changes in that other issue. Take a look if you like, I would love some feedback -- but I don't think it's ready yet for submitting here...

**Comment by erikkemperman** [comment]
@deepu105 @jdubois @atomfrede

Just to follow up on this quickly: I have tried to make a jhipster module out of this, but it is getting too ugly for my taste.

In particular I need to manipulate `pom.xml` in ways that the generator base does not facilitate: it has only `addMavenDependency` and `addMavenPlugin` but I would have to replace a dependency and add exclusion(s). Using `rewriteFile` or `replaceContent` for such things seems to me to be quite error-prone.

So in short, no hard feelings about not considering this PR for master, but my impressions is now that it doesn't really work as a module, either. So I guess I will keep my patch around for private use, but don't hold your breath for Log4J2 support in the marketplace :-)

Thanks again for your time!

**Comment by atomfrede** [comment]
Yes it might be complicated. But adding the log4j2 as dependency can be done via ``addMavenDependency`` with exclusions. Removing the logback can only be done via rewriting or replacing.

**Comment by erikkemperman** [comment]
@atomfrede Actually, it should replace the whole `spring-boot-starter-logging` artifact by `spring-boot-starter-log4j2`, and add an exclusion on the former within the (existing) `spring-boot-starter-aop` artifact.

My feeling when trying this out is that this is perhaps just a bit more involved than anticipated by the tools provided to jhipster modules.

**Comment by gmarziou** [comment]
Maybe your module execution could end by printing a link to  manual instructions about how to modify your pom.xml

**Comment by cbornet** [comment]
Removing a pom dependency should be quite simple to do with a regexp. We could even create a method for this in script-base.

**Comment by erikkemperman** [comment]
That would certainly be helpful, if this were a function in the base. If there were also a function to add an exclusion to an existing dependency (which is trickier I suspect to cover all cases) then I'd be all set!

Maybe it's just me but if "half" of this module works by providing verbal instructions to the user, it almost defeats the purpose of having a generator :-)

**Comment by gmarziou** [comment]
Rather than using a regexp to edit the pom.xml, I would try to look at  node modules to parse and edit XML files

https://github.com/matthewmatician/xml-flow

https://www.npmjs.com/package/sax-streamer

**Comment by cbornet** [comment]
Will this keep the comments ?

**Comment by gmarziou** [comment]
Good point, I don't know.

**Comment by gmarziou** [comment]
[sax-streamer](https://www.npmjs.com/package/sax-streamer) keeps comments, I just tried.
I must say that the README is a bit misleading but the unit tests are well written and really helpful.
